### PR TITLE
Feature/blog pagination v2

### LINKED ISSUE #1003: Article filter
### Link task to story

Story: https://github.com/Open-Study-College/osc/issues/1000

### Description

- Utilise button group
- Algolia powered filters (RefinementList attached to a button group)
- Factor in the Pagination / Load more.

### PR DESCRIPTION
**Pull request by jkellerman**
### 🧹 Please check off the following before submitting:

- [x] Close issues
- [x] Request reviewers
- [x] Assign yourself
- [x] Add labels (e.g. `build:osc-ecommerce`)
- [x] Add project
- [x] Set project status
- [x] Add milestone

---

## ☑ List all github issues that this PR will close:

- Closes #1003 

## 📝 Description

This is a second version of blog-pagination PR, which solves some conflict issues we we're experiences in the previous version. It also hopes to solve the issue we have with the blog page not loading in production.

It creates some new features, including:

1) Utilises the button group to allow navigation between category posts
2) Introduces pagination

## ⛳️ Current behaviour (updates)

N/A

## 🚀 New behaviour

N/A

## 💣 Is this a breaking change (Yes/No): No

> Please describe any breaking changes.

## 🧾 Final Checklist

- [ ] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [ ] I've considered possible cross-device impacts (iOS, Android)
- [ ] I've considered possible cross-browser impacts (Chrome, Safari, Edge, Firefox)
- [ ] I've added tests to cover my changes (Unit, E2E, Integration, Contract, Functional, Acceptance etc)
- [ ] I've added or updated any documentation.

## 📝 Additional Information

Expecting performance hit on blog index page due to allPosts data fetching a number of posts. Will look to solve this issue using pagination in groq query in a separate PR. For now, we need to see if the we can get the blog page working.

### Summary by CodeRabbit

**New Features:**
- Added pagination to the `PostsGrid` component, allowing users to navigate through multiple pages of blog posts.
- Enhanced `ButtonGroup` component with form submission capabilities for better user interaction.
- Introduced URL-friendly category slugs in the `categoryModule` interface, improving SEO and usability.

**Refactor:**
- Simplified data fetching in the `loader` function of the blog index route, reducing complexity and improving maintainability.
- Streamlined the retrieval of specific properties from post data in the `FeaturedCard` component, enhancing code readability.
- Consolidated multiple individual queries into a single query in the blog index route, improving performance and maintainability.

**Bug Fix:**
- Improved error handling in the `getPageData` function by rethrowing errors, enabling better error tracking and resolution.

### CODE REVIEW

File: packages/osc-ecommerce/app/routes/blog/index.tsx [hunk 1]
```diff
@@ -91,55 +66,110 @@ export const loader: LoaderFunction = async ({ request, params }) => {
...
+            request,
+            query: POST_CATEGORY,
+            params: {
+                query: query,
+                allData: 'true',
+            },
+        });
+    } else {
+        // fetches posts for all articles query param and blog index page
+        latestNewsPost = await getSettingsData({
```
Comment (stevenbister) [inline packages/osc-ecommerce/app/routes/blog/index.tsx:89]: Unfortunately This page is taking like 4-5 seconds to load on my machine :/

https://github.com/Open-Study-College/osc/assets/23461173/0cfb195e-571a-46c7-99d9-71f96885dc65

I think putting these into a promise.all will help cut down the loading time.
Something along these lines maybe?

```tsx
// Perhaps name these better lol
const [lnp, lca, lst, lss, lfp, ap] = await Promise.all([
            // fetches posts for all articles query param and blog index page
            getSettingsData({
                query: LATEST_NEWS_POST,
            }),
            getSettingsData({
                query: LATEST_CAREER_ADVICE_POST,
            }),
            getSettingsData({
                query: LATEST_STUDY_TIPS_POST,
            }),
            getSettingsData({
                query: LATEST_SUCCESS_STORIES_POST,
            }),
            getSettingsData({
                query: LATEST_FEATURED_POST,
            }),
            getSettingsData({
                query: ALL_POSTS,
                params: {
                    allData: 'true',
                },
            }),
        ]);

        latestNewsPost = lnp;
        latestCareersAdvicePost = lca;
        latestStudyTipPost = lst;
        latestSuccessStoriesPost = lss;
        latestFeaturedPost = lfp;
        allPosts = ap;
```

**Review by stevenbister — CHANGES_REQUESTED** [review]
Looks good mate, one comment around the perf but otherwise good job!

File: packages/osc-ecommerce/app/routes/blog/index.tsx (hunk 1 above)
Comment (stevenbister) [reply packages/osc-ecommerce/app/routes/blog/index.tsx:89]: I would also maybe add a todo and open a ticket to remind us to add pagination to the query in the future so we're not loading as much data, that should help too!

Comment (jkellerman) [reply packages/osc-ecommerce/app/routes/blog/index.tsx:89]: I already mentioned this in the additional notes about the performance issues and the need to use pagination in groq instead which I will sort out in a separate PR. I can add in the promise.all but i don't think it makes much difference.

Comment (stevenbister) [reply packages/osc-ecommerce/app/routes/blog/index.tsx:89]: Ah yeah sorry missed that! The promise.all did shave off 2-3 seconds for me so defo worth adding I think

Comment (jkellerman) [reply packages/osc-ecommerce/app/routes/blog/index.tsx:89]: const [lnp, lca, lst, lss, lfp, ap] = await Promise.all

THis actually might be the best way to name them haha.

**Review by stevenbister — APPROVED** [review]
LGTM!

**Pull request merged**